runtime/onert/api/include/nnfw_experimental.h

```diff
@@ -263,11 +263,33 @@ NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const vo
 NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, const void *expected,
                                     const nnfw_tensorinfo *expected_info);
 
+/**
+ * @brief Set training output buffer
+ *
+ * This function must be called after {@link nnfw_train_prepare}, \p buffer given to this function
+ * can be reused for training. \p length must be greater or equal than the operand requires.
+ * An output operand can have unspecified shape and deduced dynamically during the execution. You
+ * must provide \p buffer large enough.
+ *
+ * @param[in]   session Session from inference output is to be extracted
+ * @param[in]   index   Index of output to be set (0-indexed)
+ * @param[in]   type    Type of the output
+ * @param[out]  buffer  Raw buffer for output
+ * @param[in]   length  Size of bytes of output buffer
+ *
+ * @return      @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_output(nnfw_session *session, uint32_t index, NNFW_TYPE type,
```

Comment:
@jyoungyun What is the difference between `nnfw_set_output` and `nnfw_train_set_output` ?

Reply:
The functions are the same. However, it should be invoked after `nnfw_train_prepare` because it is changes the output shape according to the batch size.

Reply:
Thank you for answer. I've found we already have `nnfw_train_set_input` perhaps due to the same reason.

Reply:
@glistening If the same `nnfw_set_input` and `nnfw_set_output` are used in the training, the `State` class must be modified to handle both training and inference status.

And we don't need to handle batch using `nnfw_set_input_tensorinfo`. The updated batch size is already applied to the internal graph in the `nnfw_train_prepare` function. The `nnfw_train_set_output/input` function sets the user buffers to the internal graph. What I want to say in the https://github.com/Samsung/ONE/pull/12265#discussion_r1424808224 comment is that these functions should be called before `nnfw_train_prepare`, not `nnfw_prepare`. To this end, the training API is different from `nnfw_set_input/output` functions. If we change the `nnfw_set_input/output` code to check whether the `nnfw_train_prepare` has already been called, we can use the same api for training.

Anyway, introducing separate APIs for training from the inference API is our policy that we decided when we first discussed. If you think that it would be better to use the same API in training and inference, let's talk about it at our weekly meeting.